### PR TITLE
Experience-8928: Add Chromatic GHA

### LINKED ISSUE #8928: Integrate Chromatic
## Problem statement

Now that we have our Storybook configuration set up, we can add a GitHub Action to automate the Chromatic deployment.

## What you need to know

- [Chromatic Storybook](https://6419b81d4a47163c3050f14a-xtlligolml.chromatic.com/?path=/story/components-staticalert--default)

## Acceptance criteria

- [ ] We have a GHA to deploy Storybooks to Chromatic

### PR DESCRIPTION
**Pull request by stephenkao**
This changeset adds a Chromatic GHA workflow to deploy our Storybook to Chromatic.  This'll run on any `push` event on any branch, and it'll automatically approve any changes on `master.  Hopefully this'll help during design planning and review since we'll have a central component styleguide!

## Linked Issues
- Fixes #8928

### CODE REVIEW

**Comment by stephenkao** [comment]
@etanb @jpandersen87 @penny-lischer Take a look at this pull request whenever you get a chance.  Whenever we make changes to files in `frontend-react/`, the new workflow will trigger a deployment to Chromatic, and it'll update the pull request with comments:

Screenshot 2023-04-06 at 10 40 48
Screenshot 2023-04-06 at 11 14 57

Regarding keeping the Chromatic baseline up-to-date, we could automate that, but I think we can take a more manual approach and review/approve ourselves for the time being till we're certain that we want to automate that.

**Review by jpandersen87 — APPROVED** [review]

File: .github/workflows/chromatic.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      - deployment
+
+jobs:
+  chromatic-deployment:
+    name: Chromatic
+    runs-on: ubuntu-latest
+    steps:
+      - name: Find Comment
+        id: comment_find
+        uses: peter-evans/find-comment@v2
```
Comment (snesm) [inline .github/workflows/chromatic.yml:21]: We use [SHA checksums](https://devopsjournal.io/blog/2021/12/11/GitHub-Actions-Maturity-Levels) for all actions
`peter-evans/find-comment@034abe94d3191f9c89d870519735beae326f2bdb`

File: .github/workflows/chromatic.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        with:
+          issue-number: ${{ github.event.pull_request.number }}
+          comment-author: 'github-actions[bot]'
+          body-regex: '\.*Chromatic.*\gi'
+
+      - name: Create comment
+        id: comment_create
+        uses: peter-evans/create-or-update-comment@v3
```
Comment (snesm) [inline .github/workflows/chromatic.yml:29]: `peter-evans/create-or-update-comment@3383acd359705b10cb1eeef05c0e88c056ea4666
`

File: .github/workflows/chromatic.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        with:
+          comment-id: ${{ steps.comment_find.outputs.comment-id }}
+          issue-number: ${{ github.event.pull_request.number }}
+          body: |
+            Deploying branch Storybook to Chromatic...
+          edit-mode: replace
+          reactions: rocket
+          reactions-edit-mode: replace
+
+      - uses: actions/checkout@v3
```
Comment (snesm) [inline .github/workflows/chromatic.yml:39]: `actions/checkout@8f4b7f84864484a7bf31766abe9204da3cbe65b3
`

File: .github/workflows/chromatic.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        with:
+          fetch-depth: 0 # need commit history for Chromatic
+
+      - uses: actions/setup-node@v3.6.0
```
Comment (snesm) [inline .github/workflows/chromatic.yml:43]: `actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c`

File: .github/workflows/chromatic.yml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+          node-version-file: frontend-react/package.json
+
+      - name: Install dependencies
+        run: yarn
+        working-directory: frontend-react
+        shell: bash
+
+      - name: Publish to Chromatic
+        id: chromatic_publish
+        uses: chromaui/action@v1
```
Comment (snesm) [inline .github/workflows/chromatic.yml:54]: current sub-version appears to be 6.17.3
`chromaui/action@d51b84e79d164fbe8fc5bb7175695d88ddd04b72`

File: .github/workflows/chromatic.yml [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+          token: ${{ secrets.GITHUB_TOKEN }}
+          projectToken: ${{ secrets.CHROMATIC_PROJECT_TOKEN }}
+          ignoreLastBuildOnBranch: "**"
+          onlyChanged: true
+          exitZeroOnChanges: true
+          exitOnceUploaded: true
+
+      - name: Update comment
+        if: steps.comment_find.outputs.comment-id != '' || steps.comment_create.outputs.comment-id != ''
+        uses: peter-evans/create-or-update-comment@v3
```
Comment (snesm) [inline .github/workflows/chromatic.yml:66]: `peter-evans/create-or-update-comment@3383acd359705b10cb1eeef05c0e88c056ea4666
`

**Review by snesm — CHANGES_REQUESTED** [review]

File: .github/workflows/chromatic.yml (hunk 1 above)
Comment (stephenkao) [reply .github/workflows/chromatic.yml]: Thanks for the callout!  I've updated -- feel free to take a look whenever!

**Comment by jeremy-page** [comment]
How will this handle production, are we assuming master & production are identical for this? 
Is everyone OK with adding 4 minutes to every push that touches the front end?

**Review by penny-lischer — APPROVED** [review]
LGTM

**Comment by etanb** [comment]
Thinking about @jeremy-page comment, I agree that activating the workflow on EVERY push does seem like overkill, especially since the amount of time developers will be needing updates within Storybook will be a fraction of their workflow.

I think if we can make it opt-in, that would be ideal. Some ideas:

- Launch script if the commit message contains certain text like `Build Storybook` or some other keyword.
- Same idea as above, but as a github comment.
- Use the exact same logic as you have above, BUT it searchs the PRs labels to see if it has `build-storybook` or some such, and only then does it build.

**Comment by stephenkao** [comment]
> How will this handle production, are we assuming master & production are identical for this? Is everyone OK with adding 4 minutes to every push that touches the front end?

> Thinking about @jeremy-page comment, I agree that activating the workflow on EVERY push does seem like overkill, especially since the amount of time developers will be needing updates within Storybook will be a fraction of their workflow.
> 
> I think if we can make it opt-in, that would be ideal. Some ideas:
> 
> * Launch script if the commit message contains certain text like `Build Storybook` or some other keyword.
> * Same idea as above, but as a github comment.
> * Use the exact same logic as you have above, BUT it searchs the PRs labels to see if it has `build-storybook` or some such, and only then does it build.

Sure, that makes sense to me.  The reason I wanted this build to happen for every frontend change was so that it'd force us to review any visual changes that should be added to the baseline in Chromatic.  However, I could add another workflow so that it just resyncs the baseline anytime something is merged into master.

Regarding the opt-in functionality, I can use the same logic for applying GitHub labels I have in [the trial frontend labels pull request](https://github.com/CDCgov/prime-reportstream/pull/9026).  Anytime we want to rebuild the branch-specific Storybook in Chromatic, we could just remove the label and re-apply it to trigger the workflow.  How's that sound?

**Review by etanb — APPROVED** [review]
Thanks so much for tackling this!

**Comment by stephenkao** [comment]
@jeremy-page Mind taking another look at this?  I've updated it with the following changes:
- For branch-specific Chromatic deployments, we can just add the `chromatic` label
- For baseline/master deployments, the changes will be auto-approved

**Review by jeremy-page — APPROVED** [review]

**Comment by stephenkao** [comment]
@snesm Would you mind approving this whenever you get a chance?  I don't think I'm able to merge since you requested changes initially 😬

**Review by snesm — APPROVED** [review]

File: frontend-react/package.json [hunk 7]
```diff
@@ -81,8 +81,7 @@
         "yarn:show-outdated-packages": "yarn outdated",
         "run-build-dir": "yarn build:localdev:csp && yarn global add serve && serve -s build",
         "storybook": "npm-run-all -p watch-scss storybook:start",
-        "storybook:start": "start-storybook -p 6006",
```
Comment (etanb) [inline frontend-react/package.json:84]: Removing this makes `npm run storybook` error out as it relies on `storybook:start` @stephenkao

Comment (stephenkao) [reply frontend-react/package.json:84]: Whoops, will put up a fix!

Comment (stephenkao) [reply frontend-react/package.json:84]: https://github.com/CDCgov/prime-reportstream/pull/9236

**Pull request merged**